pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -148,56 +148,166 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
+			expireTimer.Reset(exp.GetExpiryFromExpirePriorityQueue())
+			exp.expirePriorityQueue.Mutex.Unlock()
 		}
 	}
-	// Send flow records to IPFIX collector.
-	err := exp.sendFlowRecords()
-	if err != nil {
-		klog.Errorf("Error when sending flow records: %v", err)
-		// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
-		// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
+}
+
+// GetExpiryFromExpirePriorityQueue returns the earliest timestamp (active expiry
+// or idle expiry) from expire priority queue.
+func (exp *flowExporter) GetExpiryFromExpirePriorityQueue() time.Duration {
+	currTime := time.Now()
+	if exp.expirePriorityQueue.Len() > 0 {
+		// Get the minExpireTime of the top item in ExpirePriorityQueue.
+		expiryDuration := MinExpiryTime + exp.expirePriorityQueue.MinExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return MinExpiryTime
+		}
+		return expiryDuration
+	}
+	if exp.activeFlowTimeout < exp.idleFlowTimeout {
+		return exp.activeFlowTimeout
 	}
-	klog.V(2).Infof("Successfully exported IPFIX flow records")
+	return exp.idleFlowTimeout
+}
+
+func (exp *flowExporter) sendFlowRecords() error {
+	exp.expirePriorityQueue.Mutex.Lock()
+	defer exp.expirePriorityQueue.Mutex.Unlock()
+
+	if exp.expirePriorityQueue.Len() == 0 {
+		return nil
+	}
+	currTime := time.Now()
+	for exp.expirePriorityQueue.Len() > 0 {
+		topItem := exp.expirePriorityQueue.Peek()
+		// If the top item is not timeout, then we do not have to check the following items.
+		if topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime) {
+			break
+		}
+		pqItem := heap.Pop(exp.expirePriorityQueue).(*flowexporter.ItemToExpire)
+		// The pqItem contains a flow record.
+		if !pqItem.IsDenyConn {
+			if pqItem.IsActive {
+				// Send a flow record if it is active and time out.
+				// TODO: more records per data set will be supported when go-ipfix supports size check when adding records
+				if err := exp.addConnToSet(pqItem.Conn, false); err != nil {
+					return err
+				}
+				if _, err := exp.sendDataSet(); err != nil {
+					return err
+				}
+				exp.numDataSetsSent = exp.numDataSetsSent + 1
+			}
+			if flowexporter.IsConnectionDying(&pqItem.Conn) {
+				// If the connection is in dying state or connection is not in conntrack
+				// table, we set the DyingAndDoneExport flag to do the deletion later.
+				pqItem.DyingAndDoneExport = true
+			} else {
+				if pqItem.IdleExpireTime.Before(currTime) && (pqItem.Conn.OriginalPackets <= pqItem.PrevPackets) && (pqItem.Conn.ReversePackets <= pqItem.PrevReversePackets) {
+					pqItem.IsActive = false
+				}
+				exp.ValidateAndUpdateStats(pqItem)
+				if pqItem.IdleExpireTime.Before(currTime) {
+					pqItem.IdleExpireTime = currTime.Add(exp.idleFlowTimeout)
+				}
```

Comment:
should we do this? 
`idleTimeout` will be reset every time we see the connection during polling.

Reply:
There are three possible alternatives:
1. we remove the connection from the PQ if it is idle time out - the problem is, in `addOrUpdateConn`, we check if there's stats change, IsActive will be set to true, idleExpireTime will be reset. Then we need to decide whether we should update the pqItem(if it still exist in the PQ), or push it back into the PQ(if it has been removed from the PQ). But we can get this info from nowhere IMO (p.s. we cannot mark `conn.pqItem = nil` when doing the remove, since we need to access `existingConn.pqItem.prevPackets` in `addOrUpdateConn`)
2. we remove the connection from the PQ if it is idle time out with no stats changes, which means inactive, then we solve the previous problem since we can check `pqItem.IsActive` field to know if it exists in PQ. But I'm wondering if that's possible an pqItem is idle time out but still active, in that case, we still need to reset its `idleExpireTime` and let it be sent multiple times, otherwise it will be pop out and push back again and again, since it is at the top of the PQ due to idle time out.
3. the third approach is the current version, we don't remove pqItem from PQ unless it is to be deleted from connection map.
Is there any suggestion? @srikartati

Reply:
Currently working on the second approach, let me push a version with it at first and then ask you to take a look.

Reply:
>we remove the connection from the PQ if it is idle time out with no stats changes, which means inactive, then we solve the previous problem since we can check pqItem.IsActive field to know if it exists in PQ. But I'm wondering if that's possible an pqItem is idle time out but still active, in that case, we still need to reset its idleExpireTime and let it be sent multiple times, otherwise it will be pop out and push back again and again, since it is at the top of the PQ due to idle time out.

I don't think there will be a case, where flow expires because of `idleTimeout` and it is active. If it is idle, we send the flow record after idle time expiry and wait for it to become active. The logic for it to become active depend on stats and tcp state. Popping out the pq item after idle expiry and pushing when it becomes active makes sense. This means you may not need the `isActive` flag in pq item anymore.

Reply:
If we don't keep `isActive`, then we don't know whether to update it in pq or push back to pq. There's a case for update because we check the stats and tcp state change for every connection, no matter their pq item exist in pq or not.